From: Lakhvich Dmitriy <LDmitriy@ru.ibm.com>
Subject: qeth: initialize net_device with carrier off
Patch-mainline: v4.5-rc1
Git-commit: e5ebe63214d44d4dcf43df02edf3613e04d671b9
References: bnc#964230, LTC#136514

Description:  qeth: layer2 reports unknown state to network tools.
Symptom:      State UNKNOWN in output of "ip link show" 
Problem:      It is caused by a missing initialization
              of the __LINK_STATE_NOCARRIER bit in the net_device state field.
Solution:     Add a netif_carrier_off() invocation when creating the
              net_device for a qeth device.
Reproduction: 1. Install linux distro.  Inspect reported state of qeth device.

Upstream-Description:

              qeth: initialize net_device with carrier off

              /sys/class/net/<interface>/operstate for an active qeth network
              interface offen shows "unknown", which translates to "state UNKNOWN
              in output of "ip link show". It is caused by a missing initialization
              of the __LINK_STATE_NOCARRIER bit in the net_device state field.
              This patch adds a netif_carrier_off() invocation when creating the
              net_device for a qeth device.

              Signed-off-by: Ursula Braun <ubraun@linux.vnet.ibm.com>
              Acked-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>
              Reference-ID: Bugzilla 133209
              Signed-off-by: David S. Miller <davem@davemloft.net>


Signed-off-by: Lakhvich Dmitriy <LDmitriy@ru.ibm.com>
Acked-by: John Jolly <jjolly@suse.com>
---
 drivers/s390/net/qeth_l2_main.c |    1 +
 drivers/s390/net/qeth_l3_main.c |    1 +
 2 files changed, 2 insertions(+)

--- a/drivers/s390/net/qeth_l2_main.c
+++ b/drivers/s390/net/qeth_l2_main.c
@@ -965,6 +965,7 @@ static int qeth_l2_setup_netdev(struct q
 	qeth_l2_request_initial_mac(card);
 	SET_NETDEV_DEV(card->dev, &card->gdev->dev);
 	netif_napi_add(card->dev, &card->napi, qeth_l2_poll, QETH_NAPI_WEIGHT);
+	netif_carrier_off(card->dev);
 	return register_netdev(card->dev);
 }
 
--- a/drivers/s390/net/qeth_l3_main.c
+++ b/drivers/s390/net/qeth_l3_main.c
@@ -3307,6 +3307,7 @@ static int qeth_l3_setup_netdev(struct q
 
 	SET_NETDEV_DEV(card->dev, &card->gdev->dev);
 	netif_napi_add(card->dev, &card->napi, qeth_l3_poll, QETH_NAPI_WEIGHT);
+	netif_carrier_off(card->dev);
 	return register_netdev(card->dev);
 }
 
